aws/resource_aws_workspaces_ip_group.go

```diff
@@ -0,0 +1,152 @@
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"rules": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"source": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsWorkspacesIpGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	rules := d.Get("rules").(*schema.Set).List()
+
+	resp, err := conn.CreateIpGroup(&workspaces.CreateIpGroupInput{
+		GroupName: aws.String(d.Get("name").(string)),
+		GroupDesc: aws.String(d.Get("description").(string)),
+		UserRules: mapIpGroupRules(rules),
+	})
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*resp.GroupId)
+
+	return resourceAwsWorkspacesIpGroupRead(d, meta)
+}
+
+func resourceAwsWorkspacesIpGroupRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	resp, err := conn.DescribeIpGroups(&workspaces.DescribeIpGroupsInput{
+		GroupIds: []*string{aws.String(d.Id())},
+	})
+	if err != nil {
+		if len(resp.Result) == 0 {
+			log.Printf("[WARN] Workspaces Ip Group (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", resp.Result[0].GroupName)
+	d.Set("description", resp.Result[0].GroupDesc)
+
+	var rules []map[string]interface{}
+
+	for _, rule := range resp.Result[0].UserRules {
+		r := map[string]interface{}{
+			"source": *rule.IpRule,
+		}
+		if rule.RuleDesc != nil {
+			r["description"] = *rule.RuleDesc
+		}
+		rules = append(rules, r)
+	}
+
+	d.Set("rules", rules)
+
+	return nil
+}
+
+func resourceAwsWorkspacesIpGroupUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	if d.HasChange("rules") {
+		rules := d.Get("rules").(*schema.Set).List()
+
+		log.Printf("[INFO] Updating Workspaces IP Group Rules")
+		_, err := conn.UpdateRulesOfIpGroup(&workspaces.UpdateRulesOfIpGroupInput{
+			GroupId:   aws.String(d.Id()),
+			UserRules: mapIpGroupRules(rules),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceAwsWorkspacesIpGroupRead(d, meta)
+}
+
+func resourceAwsWorkspacesIpGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	log.Printf("[INFO] Deleting Workspaces IP Group")
+	_, err := conn.DeleteIpGroup(&workspaces.DeleteIpGroupInput{
+		GroupId: aws.String(d.Id()),
+	})
+	if err != nil {
+		return fmt.Errorf("Error Deleting Workspaces IP Group: %s", err)
+	}
+
+	return nil
+}
+
+func mapIpGroupRules(rules []interface{}) []*workspaces.IpRuleItem {
```

Comment:
We have a pattern of naming these functions `expand...`, and having a matching `flatten...` that would replace the inline loop at line 92.

Reply:
Fixed